javax.crypto.AEADBadTagException crashes in Keystore.decryptBytes

Two crashes reported here which seem very related? Currently affect about ~dozen people.

See also #165 

```
java.lang.RuntimeException: 
  at android.app.ActivityThread.handleBindApplication (ActivityThread.java:5942)
  at android.app.ActivityThread.access$1100 (ActivityThread.java:202)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1658)
  at android.os.Handler.dispatchMessage (Handler.java:106)
  at android.os.Looper.loop (Looper.java:193)
  at android.app.ActivityThread.main (ActivityThread.java:6744)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:495)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:859)
Caused by: javax.crypto.AEADBadTagException: 
  at android.security.keystore.AndroidKeyStoreCipherSpiBase.engineDoFinal (AndroidKeyStoreCipherSpiBase.java:517)
  at javax.crypto.Cipher.doFinal (Cipher.java:2055)
  at mozilla.components.lib.dataprotect.Keystore.decryptBytes (Keystore.kt:244)
  at mozilla.lockbox.support.SecurePreferences.getString (SecurePreferences.kt:40)
  at mozilla.lockbox.store.AccountStore.getStoredAccountJSON (AccountStore.kt:76)
  at mozilla.lockbox.store.AccountStore.detectAccount (AccountStore.kt:120)
  at mozilla.lockbox.store.AccountStore.injectContext (AccountStore.kt:116)
  at mozilla.lockbox.LockboxApplication.injectContext (LockboxApplication.kt:108)
  at mozilla.lockbox.LockboxApplication.onCreate (LockboxApplication.kt:59)
  at android.app.Instrumentation.callApplicationOnCreate (Instrumentation.java:1155)
  at android.app.ActivityThread.handleBindApplication (ActivityThread.java:5937)
Caused by: android.security.KeyStoreException: 
  at android.security.KeyStore.getKeyStoreException (KeyStore.java:905)
  at android.security.keystore.KeyStoreCryptoOperationChunkedStreamer.doFinal (KeyStoreCryptoOperationChunkedStreamer.java:224)
  at android.security.keystore.AndroidKeyStoreAuthenticatedAESCipherSpi$BufferAllOutputUntilDoFinalStreamer.doFinal (AndroidKeyStoreAuthenticatedAESCipherSpi.java:373)
  at android.security.keystore.AndroidKeyStoreCipherSpiBase.engineDoFinal (AndroidKeyStoreCipherSpiBase.java:506)
```

```
java.lang.RuntimeException: 
  at android.app.ActivityThread.handleBindApplication (ActivityThread.java:5743)
  at de.robv.android.xposed.XposedBridge.invokeOriginalMethodNative (Native Method)
  at de.robv.android.xposed.XposedBridge.handleHookedMethod (XposedBridge.java:361)
  at android.app.ActivityThread.handleBindApplication (<Xposed>)
  at android.app.ActivityThread.-wrap1 (Unknown Source)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1656)
  at android.os.Handler.dispatchMessage (Handler.java:106)
  at android.os.Looper.loop (Looper.java:164)
  at android.app.ActivityThread.main (ActivityThread.java:6494)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:440)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:807)
  at de.robv.android.xposed.XposedBridge.main (XposedBridge.java:108)
Caused by: javax.crypto.AEADBadTagException: 
  at android.security.keystore.AndroidKeyStoreCipherSpiBase.engineDoFinal (AndroidKeyStoreCipherSpiBase.java:517)
  at javax.crypto.Cipher.doFinal (Cipher.java:1741)
  at mozilla.components.lib.dataprotect.Keystore.decryptBytes (Keystore.kt:244)
  at mozilla.lockbox.support.SecurePreferences.getString (SecurePreferences.kt:40)
  at mozilla.lockbox.store.AccountStore.getStoredAccountJSON (AccountStore.kt:76)
  at mozilla.lockbox.store.AccountStore.detectAccount (AccountStore.kt:120)
  at mozilla.lockbox.store.AccountStore.injectContext (AccountStore.kt:116)
  at mozilla.lockbox.LockboxApplication.injectContext (LockboxApplication.kt:108)
  at mozilla.lockbox.LockboxApplication.onCreate (LockboxApplication.kt:59)
  at android.app.Instrumentation.callApplicationOnCreate (Instrumentation.java:1119)
  at android.app.ActivityThread.handleBindApplication (ActivityThread.java:5740)
Caused by: android.security.KeyStoreException: 
  at android.security.KeyStore.getKeyStoreException (KeyStore.java:697)
  at android.security.keystore.KeyStoreCryptoOperationChunkedStreamer.doFinal (KeyStoreCryptoOperationChunkedStreamer.java:224)
  at android.security.keystore.AndroidKeyStoreAuthenticatedAESCipherSpi$BufferAllOutputUntilDoFinalStreamer.doFinal (AndroidKeyStoreAuthenticatedAESCipherSpi.java:373)
  at android.security.keystore.AndroidKeyStoreCipherSpiBase.engineDoFinal (AndroidKeyStoreCipherSpiBase.java:506)
```

--- a/.idea/codeStyles/Project.xml
+++ b/.idea/codeStyles/Project.xml
@@ -13,31 +13,6 @@
     <MarkdownNavigatorCodeStyleSettings>
       <option name="RIGHT_MARGIN" value="72" />
     </MarkdownNavigatorCodeStyleSettings>
-    <Objective-C-extensions>
-      <file>
-        <option name="com.jetbrains.cidr.lang.util.OCDeclarationKind" value="Import" />
-        <option name="com.jetbrains.cidr.lang.util.OCDeclarationKind" value="Macro" />
-        <option name="com.jetbrains.cidr.lang.util.OCDeclarationKind" value="Typedef" />
-        <option name="com.jetbrains.cidr.lang.util.OCDeclarationKind" value="Enum" />
-        <option name="com.jetbrains.cidr.lang.util.OCDeclarationKind" value="Constant" />
-        <option name="com.jetbrains.cidr.lang.util.OCDeclarationKind" value="Global" />
-        <option name="com.jetbrains.cidr.lang.util.OCDeclarationKind" value="Struct" />
-        <option name="com.jetbrains.cidr.lang.util.OCDeclarationKind" value="FunctionPredecl" />
-        <option name="com.jetbrains.cidr.lang.util.OCDeclarationKind" value="Function" />
-      </file>
-      <class>
-        <option name="com.jetbrains.cidr.lang.util.OCDeclarationKind" value="Property" />
-        <option name="com.jetbrains.cidr.lang.util.OCDeclarationKind" value="Synthesize" />
-        <option name="com.jetbrains.cidr.lang.util.OCDeclarationKind" value="InitMethod" />
-        <option name="com.jetbrains.cidr.lang.util.OCDeclarationKind" value="StaticMethod" />
-        <option name="com.jetbrains.cidr.lang.util.OCDeclarationKind" value="InstanceMethod" />
-        <option name="com.jetbrains.cidr.lang.util.OCDeclarationKind" value="DeallocMethod" />
-      </class>
-      <extensions>
-        <pair source="cpp" header="h" fileNamingConvention="NONE" />
-        <pair source="c" header="h" fileNamingConvention="NONE" />
-      </extensions>
-    </Objective-C-extensions>
     <codeStyleSettings language="kotlin">
       <option name="CODE_STYLE_DEFAULTS" value="KOTLIN_OFFICIAL" />
       <option name="KEEP_BLANK_LINES_IN_DECLARATIONS" value="1" />

--- a/.idea/encodings.xml
+++ b/.idea/encodings.xml
@@ -1,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="Encoding" addBOMForNewFiles="with NO BOM" />
+</project>

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -121,8 +121,8 @@ dependencies {
     implementation 'com.android.installreferrer:installreferrer:1.0'
     implementation "org.jetbrains.kotlin:kotlin-reflect:$kotlin_version"
     testImplementation 'junit:junit:4.12'
-    testImplementation 'org.mockito:mockito-core:2.22.0'
-    testImplementation 'org.robolectric:robolectric:4.0.2'
+    testImplementation 'org.mockito:mockito-core:2.24.5'
+    testImplementation 'org.robolectric:robolectric:4.1'
     testImplementation "org.powermock:powermock-module-junit4:2.0.0-beta.5"
     testImplementation "org.powermock:powermock-api-mockito2:2.0.0-beta.5"
     testImplementation "org.junit.jupiter:junit-jupiter-api:5.3.1"

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -88,7 +88,7 @@ dependencies {
     implementation "androidx.recyclerview:recyclerview:$recyclerview_version"
     implementation 'androidx.legacy:legacy-support-v4:1.0.0'
     implementation 'androidx.exifinterface:exifinterface:1.0.0'
-    implementation "androidx.constraintlayout:constraintlayout:2.0.0-alpha5"
+    implementation "androidx.constraintlayout:constraintlayout:${constraintlayout_version}"
     implementation "com.jakewharton.rxbinding2:rxbinding:${rxbinding_version}"
     implementation "com.jakewharton.rxbinding2:rxbinding-design-kotlin:${rxbinding_version}"
     implementation "com.jakewharton.rxbinding2:rxbinding-appcompat-v7-kotlin:${rxbinding_version}"

--- a/app/src/main/java/mozilla/lockbox/store/AccountStore.kt
+++ b/app/src/main/java/mozilla/lockbox/store/AccountStore.kt
@@ -38,7 +38,6 @@ import mozilla.lockbox.support.Constant
 import mozilla.lockbox.support.Optional
 import mozilla.lockbox.support.SecurePreferences
 import mozilla.lockbox.support.asOptional
-import java.lang.Exception
 import java.util.concurrent.TimeUnit
 import kotlin.coroutines.CoroutineContext
 import org.mozilla.fxaclient.internal.FxaException

--- a/app/src/main/java/mozilla/lockbox/store/AccountStore.kt
+++ b/app/src/main/java/mozilla/lockbox/store/AccountStore.kt
@@ -38,9 +38,10 @@ import mozilla.lockbox.support.Constant
 import mozilla.lockbox.support.Optional
 import mozilla.lockbox.support.SecurePreferences
 import mozilla.lockbox.support.asOptional
+import java.lang.Exception
 import java.util.concurrent.TimeUnit
 import kotlin.coroutines.CoroutineContext
-import org.mozilla.fxaclient.internal.FxaException as FxaException
+import org.mozilla.fxaclient.internal.FxaException
 
 @ExperimentalCoroutinesApi
 open class AccountStore(
@@ -123,7 +124,7 @@ open class AccountStore(
             } else {
                 try {
                     this.fxa = FirefoxAccount.fromJSONString(accountJSON)
-                } catch (e: FxaException) {
+                } catch (e: Exception) {
                     pushError(e)
                 }
                 generateLoginURL()
@@ -178,7 +179,7 @@ open class AccountStore(
             val config = Config.release(Constant.FxA.clientID, Constant.FxA.redirectUri)
             fxa = FirefoxAccount(config)
             generateLoginURL()
-        } catch (e: FxaException) {
+        } catch (e: Exception) {
             this.pushError(e)
         }
         (syncCredentials as Subject).onNext(Optional(null))

--- a/app/src/main/java/mozilla/lockbox/store/AccountStore.kt
+++ b/app/src/main/java/mozilla/lockbox/store/AccountStore.kt
@@ -124,7 +124,7 @@ open class AccountStore(
             } else {
                 try {
                     this.fxa = FirefoxAccount.fromJSONString(accountJSON)
-                } catch (e: Exception) {
+                } catch (e: FxaException) {
                     pushError(e)
                 }
                 generateLoginURL()
@@ -179,7 +179,7 @@ open class AccountStore(
             val config = Config.release(Constant.FxA.clientID, Constant.FxA.redirectUri)
             fxa = FirefoxAccount(config)
             generateLoginURL()
-        } catch (e: Exception) {
+        } catch (e: FxaException) {
             this.pushError(e)
         }
         (syncCredentials as Subject).onNext(Optional(null))

--- a/app/src/main/java/mozilla/lockbox/support/FxASyncDataStoreSupport.kt
+++ b/app/src/main/java/mozilla/lockbox/support/FxASyncDataStoreSupport.kt
@@ -15,7 +15,7 @@ import mozilla.lockbox.store.ContextStore
 import java.security.SecureRandom
 
 /**
- * @param keyStrength The strength ofthe generated key in bits
+ * @param keyStrength The strength of the generated key in bits
  */
 internal fun generateEncryptionKey(keyStrength: Int): String {
     val bytes = ByteArray(keyStrength / 8)

--- a/app/src/main/java/mozilla/lockbox/support/SecurePreferences.kt
+++ b/app/src/main/java/mozilla/lockbox/support/SecurePreferences.kt
@@ -14,6 +14,7 @@ import mozilla.components.lib.dataprotect.Keystore
 import mozilla.lockbox.log
 import mozilla.lockbox.store.ContextStore
 import java.nio.charset.StandardCharsets
+import java.security.GeneralSecurityException
 import javax.crypto.AEADBadTagException
 
 private const val BASE_64_FLAGS = Base64.URL_SAFE or Base64.NO_PADDING
@@ -38,16 +39,19 @@ open class SecurePreferences(
         return if (prefs.contains(key)) {
             val value = prefs.getString(key, "")
             val encrypted = Base64.decode(value, BASE_64_FLAGS)
-            try {
-                val plain = keystore.decryptBytes(encrypted)
-                String(plain, StandardCharsets.UTF_8)
+
+            val plain = try {
+                keystore.decryptBytes(encrypted)
             } catch (error: IllegalArgumentException) {
                 log.error("IllegalArgumentException exception: ", error)
                 null
-            } catch (error: AEADBadTagException) {
-                // TODO: Handle this error better once the A-C issue has been triaged: android-components/issues/2841
-                log.error("DecryptBytes exception: ", error)
+            } catch (error: GeneralSecurityException) {
+                log.error("Decrypt exception: ", error)
                 null
+            }
+
+            plain.let {
+                if (it != null) String(it, StandardCharsets.UTF_8) else null
             }
         } else {
             null

--- a/app/src/main/java/mozilla/lockbox/support/SecurePreferences.kt
+++ b/app/src/main/java/mozilla/lockbox/support/SecurePreferences.kt
@@ -15,7 +15,6 @@ import mozilla.lockbox.log
 import mozilla.lockbox.store.ContextStore
 import java.nio.charset.StandardCharsets
 import java.security.GeneralSecurityException
-import javax.crypto.AEADBadTagException
 
 private const val BASE_64_FLAGS = Base64.URL_SAFE or Base64.NO_PADDING
 

--- a/app/src/main/java/mozilla/lockbox/support/SecurePreferences.kt
+++ b/app/src/main/java/mozilla/lockbox/support/SecurePreferences.kt
@@ -14,6 +14,7 @@ import mozilla.components.lib.dataprotect.Keystore
 import mozilla.lockbox.log
 import mozilla.lockbox.store.ContextStore
 import java.nio.charset.StandardCharsets
+import javax.crypto.AEADBadTagException
 
 private const val BASE_64_FLAGS = Base64.URL_SAFE or Base64.NO_PADDING
 
@@ -40,10 +41,11 @@ open class SecurePreferences(
             try {
                 val plain = keystore.decryptBytes(encrypted)
                 String(plain, StandardCharsets.UTF_8)
-            } catch (illegalArgError: IllegalArgumentException) {
-                log.error("IllegalArgumentException exception: ", illegalArgError)
+            } catch (error: IllegalArgumentException) {
+                log.error("IllegalArgumentException exception: ", error)
                 null
-            } catch (error: Throwable) {
+            } catch (error: AEADBadTagException) {
+                // TODO: Handle this error better once the A-C issue has been triaged: android-components/issues/2841
                 log.error("DecryptBytes exception: ", error)
                 null
             }

--- a/app/src/main/java/mozilla/lockbox/support/SecurePreferences.kt
+++ b/app/src/main/java/mozilla/lockbox/support/SecurePreferences.kt
@@ -11,6 +11,7 @@ import android.content.SharedPreferences
 import android.preference.PreferenceManager
 import android.util.Base64
 import mozilla.components.lib.dataprotect.Keystore
+import mozilla.lockbox.LogProvider.Companion.log
 import mozilla.lockbox.store.ContextStore
 import java.nio.charset.StandardCharsets
 
@@ -39,7 +40,8 @@ open class SecurePreferences(
             try {
                 val plain = keystore.decryptBytes(encrypted)
                 String(plain, StandardCharsets.UTF_8)
-            } catch (error: IllegalArgumentException) {
+            } catch (error: Throwable) {
+                log.error("DecryptBytes exception: ", error)
                 null
             }
         } else {

--- a/app/src/main/java/mozilla/lockbox/support/SecurePreferences.kt
+++ b/app/src/main/java/mozilla/lockbox/support/SecurePreferences.kt
@@ -11,7 +11,7 @@ import android.content.SharedPreferences
 import android.preference.PreferenceManager
 import android.util.Base64
 import mozilla.components.lib.dataprotect.Keystore
-import mozilla.lockbox.LogProvider.Companion.log
+import mozilla.lockbox.log
 import mozilla.lockbox.store.ContextStore
 import java.nio.charset.StandardCharsets
 
@@ -40,6 +40,9 @@ open class SecurePreferences(
             try {
                 val plain = keystore.decryptBytes(encrypted)
                 String(plain, StandardCharsets.UTF_8)
+            } catch (illegalArgError: IllegalArgumentException) {
+                log.error("IllegalArgumentException exception: ", illegalArgError)
+                null
             } catch (error: Throwable) {
                 log.error("DecryptBytes exception: ", error)
                 null

--- a/build.gradle
+++ b/build.gradle
@@ -18,6 +18,7 @@ buildscript {
     ext.picasso_version = '2.71828'
     ext.recyclerview_version = '1.1.0-alpha04'
     ext.sentry_version = '1.7.16'
+    ext.constraintlayout_version = '2.0.0-alpha5'
 
     repositories {
         google()

